sensor: scd30: Add initial support for the SCD30 sensor

burumaj opened the pull request:
This pull request add support for the scd30 sensor. Most of the code has been taken from the Sensirion github page and has been modified to work with Zephyr. The Sensirion Code is licensed as BSD 3-Clause. 
Before merging please let me know if I should change the License identifier and the copyright notice>

Fixes #39234

tejlmand commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -68,6 +68,7 @@ add_subdirectory_ifdef(CONFIG_QDEC_NRFX		qdec_nrfx)
 add_subdirectory_ifdef(CONFIG_QDEC_SAM		qdec_sam)
 add_subdirectory_ifdef(CONFIG_TEMP_NRF5		nrf5)
 add_subdirectory_ifdef(CONFIG_SBS_GAUGE		sbs_gauge)
+add_subdirectory_ifdef(CONFIG_SCD30         scd30)
```
although we generally use spaces in CMake files, then this file seems to have everything aligned with tabs, so please follow that principle so it looks similar to all other lines:
```suggestion
add_subdirectory_ifdef(CONFIG_SCD30<tab><tab>scd30)
```

tejlmand replied:
seems you added 3 tabs, not 2.

small note, please let the reviewer decide if the comment is resolved :wink:

burumaj replied:
Should be fixed now

tejlmand commented on drivers/sensor/scd30/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+zephyr_library()
+zephyr_library_sources(scd30.c)
```
missing newline.

tejlmand commented on drivers/sensor/scd30/Kconfig:
```diff
@@ -0,0 +1,30 @@
+menuconfig SCD30
+    depends on I2C
+    select SENSIRION_COMMON
+    bool "Sensirion SCD30 CO2 Sensor"
```
Please move this line up.

Makes it easier to see the type and prompt text of this setting.

tejlmand commented on drivers/sensor/scd30/Kconfig:
```diff
@@ -0,0 +1,30 @@
+    depends on I2C
+    select SENSIRION_COMMON
+    bool "Sensirion SCD30 CO2 Sensor"
+    default false
```
Please remove:
```suggestion
```

Ref: https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#redundant-defaults

tejlmand commented on drivers/sensor/scd30/Kconfig:
```diff
@@ -0,0 +1,30 @@
+
+if SCD30
+
+comment "SCD30 Options"
```
Why this ?

You are already in a submenu and can see in the top navigation bar in menuconfig under which menu you are currently.

tejlmand commented on drivers/sensor/scd30/Kconfig:
```diff
@@ -0,0 +1,30 @@
+      Has to be in the range [2; 1800].
+
+config SCD30_ENABLE_ALTITUDE_COMPENSATION
+    bool "Enable altitude compensation"
```
As per the guidelines, please remove `Enable`:
https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#prompt-strings

> For a Kconfig symbol that enables a driver/subsystem FOO, consider having just “Foo” as the prompt, instead of “Enable Foo support” or the like.

Also help text is missing.
I can guess what `altitude compensation` means, but what I cannot guess is in which way the measurement is compensated.

I assume it's the CO2 measurement that is compensated, but in which way and what does that mean.

tejlmand commented on dts/bindings/sensor/sensirion,scd30.yaml:
```diff
@@ -0,0 +1,9 @@
+
+properties:
+    label:
+      required: true
```
missing newline

tejlmand commented on drivers/sensor/scd30/scd30.h:
```diff
@@ -0,0 +1,59 @@
+int scd30_check_crc(uint8_t *data, uint8_t data_len, uint8_t checksum);
+uint8_t scd30_generate_crc(uint8_t *data, uint8_t data_len);
+
+#endif /* ZEPHYR_DRIVER_SENSOR_SCD30_SCD30_H_ */
```
missing newline

tejlmand commented on drivers/sensor/scd30/CMakeLists.txt:
```diff
@@ -1,2 +1,4 @@
 zephyr_library()
-zephyr_library_sources(scd30.c)
\ No newline at end of file
+zephyr_library_sources_ifdef(CONFIG_SCD30 scd30.c)
```
missing newline

tejlmand replied:
Why the `ifdef()` version here ?

You already safeguarded inclusion of this file like: `add_subdirectory_ifdef(CONFIG_SCD30 scd30)` in previous commit.

tejlmand commented on drivers/sensor/scd30/Kconfig:
```diff
@@ -17,12 +21,6 @@ config SCD30_MEASUREMENT_INTERVAL
-
-config SCD30_ENABLE_OFFSETS
-    bool "Enable temperature and RH offsets"
-
```
you just added those in previous commit, and now you remove them again.

Please cleanup your commits by squashing the commits.

tejlmand reviewed (changes requested):
Thanks for your contribution.

I have some comments related to the build system code.

Also we need a test or sample that uses this sensor so we can ensure that code is building correctly.

burumaj commented:
I think I have resolved all comments. I have also rebased on main to solve some merge conflict and added a sample application.

tejlmand commented on drivers/sensor/scd30/Kconfig:
```diff
@@ -0,0 +1,24 @@
+if SCD30
+
+config SCD30_MEASUREMENT_INTERVAL
+	int "measurement interval in seconds"
```
Minor nit:

```suggestion
	int "Measurement interval in seconds"
```

mbolivar-nordic commented on samples/sensor/scd30/src/main.c:
```diff
@@ -0,0 +1,44 @@
+
+void main(void)
+{
+	const struct device *dev = device_get_binding("SCD30_0");
```
Please use DEVICE_DT_GET_ANY or DEVICE_DT_GET_ONE. Make sure to check for readiness with device_is_ready.

mbolivar-nordic commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,267 @@
+	return rc;
+}
+
+DEVICE_DT_INST_DEFINE(0, scd30_init, device_pm_control_nop, &scd30_0_data, &scd30_0_cfg,
```
Please:

- use i2c_dt_spec for managing the bus device and address (and use device_is_ready properly in the init function to check the bus is ready before using it)
- support multiple instances using DT_INST_FOREACH_STATUS_OKAY

mbolivar-nordic commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,267 @@
+
+	if (i2c == NULL) {
+		LOG_DBG("Failed to get pointer to %s device!", log_strdup(cfg->bus_name));
+		return -EINVAL;
```
Please use ENODEV for unready devices

mbolivar-nordic commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,267 @@
+		return -EIO;
+	}
+
+	rc = i2c_read(scd30_i2c_device(dev), rx_buf, SCD30_WORD_SIZE + SCD30_CRC8_LEN,
```
Please use the `_dt` helpers to avoid boilerplate, e.g. use `i2c_read_dt` here instead.

mbolivar-nordic reviewed (changes requested):
A few comments from a partial review.

burumaj commented:
The latest push should fix all comments that are open. It also removes `scd30_bytes_to_uint*()` functions in favor of those
found in `sys/byteorder.h`.

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,241 @@
+#include <sys/byteorder.h>
+#include <sys/crc.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(SCD30, CONFIG_SCD30_LOG_LEVEL);
```
`SENSOR` log level?

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,241 @@
+	rc = i2c_read_dt(&cfg->bus, rx_buf, sizeof(rx_buf));
+	if (rc != 0) {
+		LOG_DBG("Failed to read data. (rc = %d)", rc);
+		return -EIO;
+	}
+
+	/* C02 data */
+	rc = scd30_fill_data_buf(rx_buf, data_buf);
+	if (rc != 0) {
+		return -EIO;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 3, data_buf + 2);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	/* temperature data */
+	rc = scd30_fill_data_buf(rx_buf + 6, data_buf + 4);
+	if (rc != 0) {
+		return -EIO;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 9, data_buf + 6);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	/* relative humidity */
+	rc = scd30_fill_data_buf(rx_buf + 12, data_buf + 8);
+	if (rc != 0) {
+		return -EIO;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 15, data_buf + 10);
+	if (rc != 0) {
+		return -EIO;
```
please, propagate error codes instead of converting to `-EIO`.

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,241 @@
+		return -EIO;
+	}
+
+	float co2_ppm, temp, rel_hum;
```
by convention: declare variables at the top

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,241 @@
+
+	float co2_ppm, temp, rel_hum;
+
+	co2_ppm = scd30_bytes_to_float(data_buf);
+	temp = scd30_bytes_to_float(data_buf + 4);
+	rel_hum = scd30_bytes_to_float(data_buf + 8);
+
+	sensor_value_from_double(&data->co2_ppm, co2_ppm);
+	sensor_value_from_double(&data->temp, temp);
+	sensor_value_from_double(&data->rel_hum, rel_hum);
```
suggestion: conversions can be done in `get`

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,241 @@
+	return tmp.float32;
+}
+
+static int scd30_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	uint8_t rx_buf[6 * (SCD30_WORD_SIZE + SCD30_CRC8_LEN)] = {};
+	uint8_t data_buf[6 * SCD30_WORD_SIZE] = {};
+	struct scd30_data *data = dev->data;
+	const struct scd30_config *cfg = dev->config;
+	int rc = 0;
+
+	rc = scd30_write_command(dev, SCD30_CMD_GET_DATA_READY);
+	if (rc != 0) {
```
this function should check if requested channel is supported (or `SENSOR_CHAN_ALL`)

burumaj replied:
One question here. If the sensor itself only supports reading all data at once, what would be the correct way of handling this? 
1) regardless of which channel is asked for, updating all channels or,
2) drop data which is not asked for and return `ENODATA` when the user asks to fetch a different channel within the sample period of the sensor.
After my latest the driver does the first option. I couldn't find the preferred way of handling this in the docs.

gmarull replied:
you may take this as an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/ina23x/ina23x.c#L156

burumaj replied:
The difference between the two sensors here, as far as I can tell, is that the ina23x supports reads for the separate channels it measures, so performing a call like `sensor_sample_fetch_chan(dev, SENSOR_CHAN_VOLTAGE)` leaves the data
for the other two channels intact in the sensor so they can be read a later stage.
So code that looks like the below snippet, will work as expected, i.e. it will update the voltage and current returned by
`sensor_channel_get()`, while leaving power untouched.
```c
sensor_sample_fetch_chan(dev, SENSOR_CHAN_VOLTAGE);
sensor_sample_fetch_chan(dev, SENSOR_CHAN_CURRENT);
```

For the SCD30 this is not the case, the sensor only supports a readout of all measurement registers at once.
This leads to the situation where perform the following two calls leads to behavior that is unexpected.
```c
sensor_sample_fetch_chan(dev, SENSOR_CHAN_CO2);
sensor_sample_fetch_chan(dev, SENSOR_CHAN_AMBIENT_TEMP);
```
The way I read the [datasheet](https://cdn.sparkfun.com/assets/d/c/0/7/2/SCD30_Interface_Description.pdf) (sections 1.3.3 and 1.3.4) this would cause the second call to `sensor_sample_fetch_chan()` to fail because no new data is available.

gmarull replied:
Thanks for the detailed explanation, in such case, you should likely validate that `chan == SENSOR_CHAN_ALL`.

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,241 @@
+	uint8_t data_buf[6 * SCD30_WORD_SIZE] = {};
+	struct scd30_data *data = dev->data;
+	const struct scd30_config *cfg = dev->config;
+	int rc = 0;
```
nit: redundant initialization

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,241 @@
+	uint8_t actual_crc = scd30_compute_crc(data, data_len);
+
+	if (checksum != actual_crc) {
+		LOG_DBG("CRC check failed");
```
`LOG_ERR`?

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,241 @@
+		return -ENODEV;
+	}
+
+	if (!cfg->bus.addr) {
+		LOG_DBG("No I2C address");
+		return -EINVAL;
+	}
```
when does this condition occur?

burumaj replied:
You are right. When an address is not specified, a compiler error is thrown. So this is redundant

gmarull commented on drivers/sensor/scd30/scd30.h:
```diff
@@ -0,0 +1,58 @@
+#include <kernel.h>
+#include <drivers/gpio.h>
+
+#define SCD30_CMD_START_PERIODIC_MEASUREMENT 0x0010
+#define SCD30_CMD_STOP_PERIODIC_MEASUREMENT 0x0104
+#define SCD30_CMD_READ_MEASUREMENT 0x0300
+#define SCD30_CMD_SET_MEASUREMENT_INTERVAL 0x4600
+#define SCD30_CMD_GET_DATA_READY 0x0202
+#define SCD30_CMD_SET_TEMPERATURE_OFFSET 0x5403
+#define SCD30_CMD_SET_ALTITUDE 0x5102
+#define SCD30_CMD_SET_FORCED_RECALIBRATION 0x5204
+#define SCD30_CMD_AUTO_SELF_CALIBRATION 0x5306
+#define SCD30_CMD_READ_SERIAL 0xD033
+#define SCD30_SERIAL_NUM_WORDS 16
+#define SCD30_WRITE_DELAY_US 20000
+
+#define SCD30_MEASUREMENT_DATA_WORDS 6
+
+#define SCD30_CRC8_POLYNOMIAL 0x31
+#define SCD30_CRC8_INIT 0xFF
+#define SCD30_CRC8_LEN 1
+
+#define SCD30_WORD_SIZE 2
+#define SCD30_COMMAND_SIZE 2
+
+#define SCD30_MAX_BUFFER_WORDS 24
+#define SCD30_CMD_SINGLE_WORD_BUF_LEN \
```
nit: add `U` suffix to unsigned constants

gmarull commented on drivers/sensor/scd30/scd30.h:
```diff
@@ -0,0 +1,58 @@
+	struct sensor_value rel_hum;
+};
+
+int scd30_write_command(const struct device *dev, uint16_t cmd);
+int scd30_write_register(const struct device *dev, uint16_t cmd, uint16_t val);
+
+int scd30_check_crc(uint8_t *data, uint8_t data_len, uint8_t checksum);
+uint8_t scd30_generate_crc(uint8_t *data, uint8_t data_len);
```
are these functions required in any other compilation unit? why are they declared here?

MaureenHelm commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2021 Advanced Climate Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
> Most of the code has been taken from the Sensirion github page and has been modified to work with Zephyr. The Sensirion Code is licensed as BSD 3-Clause.

You need to attribute this source, and you cannot re-license it on your own.

MaureenHelm commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,239 @@
+
+#include "scd30.h"
+
+int scd30_write_command(const struct device *dev, uint16_t cmd)
```
Please make all functions in this driver `static`

MaureenHelm commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	/* delay for 3 msec as per datasheet. */
+	k_busy_wait(3000);
```
This should be `k_sleep` to allow other threads to do work while we're waiting.

MaureenHelm commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	/* C02 data */
+	rc = scd30_fill_data_buf(rx_buf, data_buf);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 3, data_buf + 2);
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* temperature data */
+	rc = scd30_fill_data_buf(rx_buf + 6, data_buf + 4);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 9, data_buf + 6);
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* relative humidity */
+	rc = scd30_fill_data_buf(rx_buf + 12, data_buf + 8);
+	if (rc != 0) {
+		return rc;
+	}
+	rc = scd30_fill_data_buf(rx_buf + 15, data_buf + 10);
+	if (rc != 0) {
+		return rc;
+	}
+
+	data->co2_ppm = scd30_bytes_to_float(data_buf);
+	data->temp = scd30_bytes_to_float(data_buf + 4);
+	data->rel_hum = scd30_bytes_to_float(data_buf + 8);
```
There are too many hard-coded offsets here. Can you make `rx_buf` and `data_buf` structures instead?

MaureenHelm reviewed (changes requested):
Please move the sample to a separate commit (can stay in the same PR). Also, please add this sensor to `tests/drivers/build_all/sensor`

mbolivar-nordic commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,284 @@
+	const struct scd30_config *cfg = dev->config;
+
+	if (!device_is_ready(cfg->bus.bus)) {
+		LOG_DBG("Failed to get pointer to %s device!", log_strdup(cfg->bus.bus->name));
```
You don't need to `log_stdrup` the name; it's in flash.

mbolivar-nordic reviewed (approved):
my comments are addressed; DT bits look OK

gmarull commented on drivers/sensor/scd30/Kconfig:
```diff
@@ -0,0 +1,24 @@
+
+if SCD30
+
+config SCD30_MEASUREMENT_INTERVAL
+	int "Measurement interval in seconds"
+	default 30
+	range 2 1800
+	help
+	  The interval with wich measurement are done.
+	  Has to be in the range [2; 1800].
```
question: do you think making this a sensor attribute (ie it can be set at runtime) would be better?

burumaj replied:
You're right. It should be added now. The sample has also been updated to reflect these changes.

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,375 @@
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+	{
+		double frequency = sensor_value_to_double(val);
+		double sample_time = 1.0 /  frequency;
```
note: consider introducing `SENSOR_ATTR_SAMPLING_PERIOD` (sometimes it is easier to specify period instead of frequency).

burumaj replied:
What would be the preferred location to do this? In a sensor specific header file i.e. in `include/sensor/scd30.h` or 
adding it to `include/sensor.h`?

gmarull replied:
I think `include/sensor.h` , since it's a quite universal attribute

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,375 @@
+	.attr_set = scd30_attr_set,
+};
+
+static int scd30_init(const struct device *dev)
```
suggestion: initial sampling period could be set in Devicetree and initialized here.

gmarull reviewed (commented):
see my comments

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -358,13 +359,20 @@ static int scd30_init(const struct device *dev)
 #define SCD30_DEFINE(inst)									\
-	static struct scd30_data scd30_data_##inst = {};					\
+	static struct scd30_data scd30_data_##inst = {						\
+		.sample_time = DT_PROP(DT_INST(inst, DT_DRV_COMPAT), sample_period)		\
```
```suggestion
		.sample_time = DT_INST_PROP(inst, sample_period)		\
```

gmarull commented on dts/bindings/sensor/sensirion,scd30.yaml:
```diff
@@ -9,3 +9,11 @@ include: i2c-device.yaml
+    sample-period:
+      type: int
+      required: false
+      default: 30
+      description: |
+        The initial sample period in seconds. The sensor will be
+        configured with this value on start up. This values must
+        be in the rang [2; 1800]. This value be changed at run time.
```
default needs to be justified, see https://docs.zephyrproject.org/latest/guides/dts/bindings.html#dt-bindings-default
scd30 seems to use `2` as default

burumaj replied:
The default of 2 seconds is only a valid when the sensor is first powered on of from the factory. When the user of the sensor
changes this values it is stored non volatile memory. This is what the [datasheet](https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/9.5_CO2/Sensirion_CO2_Sensors_SCD30_Interface_Description.pdf) says in chapter 1.4.3.:

> Sets the interval used by the SCD30 sensor to measure in continuous measurement mode (see chapter 1.4.1). Initial value is  2 s. The chosen measurement interval is saved in non-volatile memory and thus is not reset to its initial value after power up.

I think having a default here is valuable, but maybe changing it to 2 is a better value since it comes from the factory this way.

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -358,13 +359,20 @@ static int scd30_init(const struct device *dev)
 
 #define SCD30_DEFINE(inst)									\
-	static struct scd30_data scd30_data_##inst = {};					\
+	static struct scd30_data scd30_data_##inst = {						\
```
use config to store const data, not data

gmarull commented:
@burumaj can you squash into:

- commit that adds new attribute to sensor.h
- commit that adds the driver
- commit that adds sample

burumaj committed:
drivers: sensor: Introduce SENSOR_ATTR_SAMPLE_PERIOD

Some sensors have use a sampling period instead of setting
a sample frequency. To more closely follow the way some
sensors expact data introduce this a sensor attribute.

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

gmarull commented on dts/bindings/sensor/sensirion,scd30.yaml:
```diff
@@ -0,0 +1,19 @@
+    sample-period:
+      type: int
+      required: false
+      default: 30
```
nit: `2`

gmarull commented on drivers/sensor/scd30/scd30.c:
```diff
@@ -0,0 +1,405 @@
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
```
do you think it's worth keeping support for this one ?

burumaj replied:
I think it is. But maybe a different RP could introduce utility function to convert from a period to a fruquency,
just like `sensor_radians_to_degrees`

burumaj committed:
drivers: sensor: scd30: Add SCD30 CO2 sensor support

Add support for the SCD30 C02 sensor in polling mode.
Only I2C is supported.

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

burumaj committed:
samples: sensor: scd30: Add scd30 sample application

Add a sample application that shows how to use the SCD30
in polled mode over I2C.

Signed-off-by: Joep Buruma <burumaj50@gmail.com>

gmarull commented on drivers/sensor/scd30/scd30.h:
```diff
@@ -0,0 +1,88 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * SPDX-License-Identifier: BSD-3-CLAUSE
```
As far as I know, in-tree code needs to be Apache-2.0.

gmarull commented on drivers/sensor/scd30/scd30.h:
```diff
@@ -0,0 +1,88 @@
+#ifndef ZEPHYR_DRIVER_SENSOR_SCD30_SCD30_H_
+#define ZEPHYR_DRIVER_SENSOR_SCD30_SCD30_H_
+
+#include <device.h>
+#include <kernel.h>
+#include <drivers/gpio.h>
```
this file only requires:
```c
#include <stdint.h>
#include <drivers/i2c.h>
```

nobodyguy commented:
For anyone looking for an updated SCD30 driver - https://github.com/nobodyguy/sensirion_zephyr_drivers
